src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -1224,47 +1224,53 @@ protected void GenerateFindFirstChar()
                 Ldc(_bmPrefix.HighASCII - _bmPrefix.LowASCII);
                 Bgtun(lDefaultAdvance);
 
-                var table = new Label[_bmPrefix.HighASCII - _bmPrefix.LowASCII + 1];
-
-                // Mapping from negative ASCII value to the label that loads it.
-                // As we create labels, we check to see if the table already has
-                // the value, and only create the label if it doesn't.  Then when
-                // spitting out the code for each label, we try to remove the entry
-                // from the dictionary, and only if we're successful (because it
-                // wasn't already removed) do we spit the code.
-                var labelMap = new Dictionary<int, Label>();
-
-                for (int i = _bmPrefix.LowASCII; i <= _bmPrefix.HighASCII; i++)
+                int negativeRange = _bmPrefix.HighASCII - _bmPrefix.LowASCII + 1;
+                if (negativeRange > 1)
                 {
-                    Label label;
-                    if (_bmPrefix.NegativeASCII[i] == beforefirst)
+                    // Create a string to store the lookup table we use to find the offset.
+                    Debug.Assert(_bmPrefix.Pattern.Length <= char.MaxValue, "RegexBoyerMoore should have limited the size allowed.");
+                    string negativeLookup = string.Create(negativeRange, (thisRef: this, beforefirst), (span, state) =>
                     {
-                        label = lDefaultAdvance;
-                    }
-                    else if (!labelMap.TryGetValue(_bmPrefix.NegativeASCII[i], out label))
+                        // Store the offsets into the string.  RightToLeft has negative offsets, so to support it with chars (unsigned), we negate
+                        // the values to be stored in the string, and then at run time after looking up the offset in the string, negate it again.
+                        for (int i = 0; i < span.Length; i++)
+                        {
+                            int offset = state.thisRef._bmPrefix!.NegativeASCII[i + state.thisRef._bmPrefix.LowASCII];
+                            if (offset == state.beforefirst)
+                            {
+                                offset = state.thisRef._bmPrefix.Pattern.Length;
+                            }
+                            else if (state.thisRef._code!.RightToLeft)
+                            {
+                                offset = -offset;
```

Comment:
I'm assuming we have enough RTL tests to find bugs in this code around here.

Reply:
I believe so.